Add support for reading items from .debug_macinfo

DanielT opened the pull request:
The section .debug_macinfo was specified in DWARF standards v2, v3 and v4. It was replaced by .debug_macro in DWARF v5. The content of .debug_macinfo is represented through the enum DebugMacInfoItem. It has variants for each specified type:
- macro defined
- macro undefined
- file includd
- end of include
- vendor extension
- end of unit

This change does not implement writing of .debug_macinfo.

philipc commented:
Thanks for working on this. I haven't reviewed your changes, but I wanted to do a quick comparision with the llvm output so I tried running dwarfdump on a file here and I get output like:
```
...
                    DW_AT_high_pc               <offset-from-lowpc>348
                    DW_AT_macro_info            Macro info: Start include file on line 0:  /object/testfiles/dwarf/base.cpp< 1><0x0000002e>    DW_TAG_variable
                      DW_AT_name                  sarr
...
```

For reference, llvm-dwarfdump simply gives the offset and lists the macinfo contents elsewhere:
```
...
              DW_AT_high_pc     (0x000000000000128c)
              DW_AT_macro_info  (0x00000000)

0x0000002e:   DW_TAG_variable
                DW_AT_name      ("sarr")
...

.debug_macinfo contents:
0x00000000:
DW_MACINFO_start_file - lineno: 0 filenum: 1
  DW_MACINFO_define - lineno: 1 macro: SIZE 10000
DW_MACINFO_end_file
DW_MACINFO_define - lineno: 0 macro: __llvm__ 1
DW_MACINFO_define - lineno: 0 macro: __clang__ 1
...
```

Note that in addition to missing a line break, the gimli output only gives one line of the macinfo data, whereas there are many more lines that would be useful to see. So I think the way llvm-dwarfdump does it is better. It's also better to simply list the constants and arguments (such as `DW_MACINFO_start_file`) instead of converting it to a description (such as "Start include file").

DanielT commented:
I just updated my branch:
Your comment made me realize that I had misunderstood the spec, and was missing most of the macinfo entries.
The DW_AT_macro_info attribute actually points to a list of items which is terminated with an end of unit item.

As a result, i changed `get_macinfo(offset)` to return an iterator over the items found starting at the `offset`.
I also modified dwarfdump to use this iterator to print all of the items. Unlike llvm-dwarfdump, they are printed inline with the DIEs, but the formatting should be generally improved compared to my first version..

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+                finished: false,
+            }
+        } else {
+            // If the offset is invalid, the iterator is immediately finished.
```
Return the error instead of an empty iterator.

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+/// an Entry in the `.debug_macinfo` section.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DebugMacInfoItem<R> {
+    /// The end of the unit.
+    EndOfUnit,
```
```suggestion
```
This is never created, and shouldn't be, so can delete this.

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+    Define {
+        /// The line number where the macro is defined.
+        line_number: u64,
+        /// The text of the macro: a null-terminated string.
```
I think we should make it clear that the text is the macro name, arguments, and value.

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+    Undef {
+        /// The line number where the macro is undefined.
+        line_number: u64,
+        /// The text of the macro: a null-terminated string.
```
Similarly this is just the macro name

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+            return None;
+        }
+        // Read the next entry from the input reader and return it as a DebugMacInfoItem.
+        let macinfo_type = DwMacInfo(self.input_reader.read_u8().ok()?);
```
Please make this a fallible iterator instead. Never do `.ok()?`.

DanielT replied:
All uses of fallible iterator depend on the "fallible-iterator" feature.
If I make this iterator a FallibleIterator, then reading the content of .debug_macinfo as a whole would have to depend on this feature. I avoided this dependency by invalidating the iterator on error, but if you think making it fallible is better then I'll change it.

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+///
+/// the iterator is not public and does not appear in the API - callers only get an `impl Iterator<Item = DebugMacInfoItem<R>>`
+struct DebugMacInfoIterator<R: Reader> {
+    input_reader: R,
```
```suggestion
    input: R,
```

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+/// the iterator is not public and does not appear in the API - callers only get an `impl Iterator<Item = DebugMacInfoItem<R>>`
+struct DebugMacInfoIterator<R: Reader> {
+    input_reader: R,
+    finished: bool,
```
```suggestion
```
Use `input.is_empty()` and `input.empty()` instead.

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+            _ => {
+                // unknown macinfo type; return None, since Err is not an option here
+                self.finished = true;
+                return None;
```
Change to return an error.

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+        // Test getting macinfo entries
+        let entry = iter.next().unwrap();
+        assert!(
+            matches!(entry, DebugMacInfoItem::Define { line_number, macro_text } if line_number == 0 && macro_text.to_string_lossy() == "__STDC__ 1")
```
```suggestion
            matches!(entry, DebugMacInfoItem::Define { line_number: 0, macro_text } if macro_text.slice() == b"__STDC__ 1")
```
Similarly for others.

philipc commented on src/read/macinfo.rs:
```diff
@@ -0,0 +1,290 @@
+    EndFile,
+    /// A vendor-specific extension.
+    VendorExt {
+        /// a numeric constant (LEB128), whose meaning is vendor specific.
+        numeric_constant: u64,
+        /// a null-terminated string, whose meaning is vendor specific.
+        string_constant: R,
```
```suggestion
        /// A numeric constant whose meaning is vendor specific.
        numeric: u64,
        /// A string whose meaning is vendor specific.
        string: R,
```
Field names are needlessly verbose. The DWARF encoding (LEB128 and null-terminated) is irrelevant here. Capitalize the start of comments.

Similarly for other variants.

philipc commented on crates/examples/src/bin/dwarfdump.rs:
```diff
@@ -1313,6 +1313,7 @@ fn dump_attr_value<R: Reader, W: Write>(
         }
         gimli::AttributeValue::DebugMacinfoRef(offset) => {
             writeln!(w, "<.debug_macinfo+0x{:08x}>", offset.0)?;
+            dump_macinfo_list(w, unit, offset)?;
```
I still think that macros can be too verbose to be listed inline like this.

DanielT commented on src/read/mod.rs:
```diff
@@ -252,6 +252,11 @@ pub use self::rnglists::*;
+#[cfg(feature = "read")]
+mod macinfo;
+#[cfg(feature = "read")]
+pub use self::macinfo::*;
```
`macro` is actually a [reserved keyword](https://doc.rust-lang.org/reference/keywords.html#reserved-keywords) in rust.

How about `debug_macro` or `macros` instead?

DanielT replied:
renamed to `macros`

DanielT commented:
New update - as far as I can tell I've addressed all of your comments.

The module is renamed to `macros` since `macro` is reserved.
I also switched to a FallibleIterator as you requested. This resulted in the addition of several `#[cfg(feature = "fallible-iterator")]` guards, so .debug_macinfo is not usable without this feature. That doesn't bother me much though, since I've never disabled this anyway.

philipc commented:
The style used in this crate is to always expose a `next` method on the struct that isn't part of any trait implementation, and then use the `fallible-iterator` feature to control the existence of only the `FallibleIterator` implementation. The `next` method in that trait implementation is simply a one liner to call the other `next` method (they both have the same signature).

DanielT commented:
Oh, I see. In that case, what is the benefit of the fallible-iterator crate?

philipc commented:
The `FallibleIterator` trait provides many more methods that may be useful for working with the iterator, such as for chaining adaptors. I personally rarely use them, but occasionally they can make code simpler.

DanielT commented:
In case you missed it - my last push yesterday changed the iterator implementation, so fallible-iterator is no longer required.

philipc commented on src/read/macros.rs:
```diff
@@ -0,0 +1,289 @@
+    /// A macro definition.
+    Define {
+        /// The line number where the macro is defined.
+        line_number: u64,
+        /// The text of the macro: The name of the macro followed immediately by any formal
+        /// parameters including the surrounding parentheses, followed by the macro definition.
+        macro_text: R,
+    },
+    /// A macro undefinition.
+    Undef {
+        /// The line number where the macro is undefined.
+        line_number: u64,
+        /// The name of the macro without the definition.
+        macro_name: R,
+    },
+    /// The start of a file.
+    StartFile {
+        /// Line number of the source file on which the inclusion macro directive occurred.
+        line_number: u64,
+        /// An index into the line number table of the compilation unit.
+        file_number_index: u64,
```
```suggestion
        line: u64,
        /// The text of the macro: The name of the macro followed immediately by any formal
        /// parameters including the surrounding parentheses, followed by the macro definition.
        text: R,
    },
    /// A macro undefinition.
    Undef {
        /// The line number where the macro is undefined.
        line: u64,
        /// The name of the macro without the definition.
        name: R,
    },
    /// The start of a file.
    StartFile {
        /// Line number of the source file on which the inclusion macro directive occurred.
        line: u64,
        /// An index into the line number table of the compilation unit.
        file: u64,
```

philipc commented on src/read/macros.rs:
```diff
@@ -0,0 +1,289 @@
+                }))
+            }
+            _ => {
+                // unknown macinfo type; return None, since Err is not an option here
```
```suggestion
```

DanielT commented:
Done!

philipc commented on crates/examples/src/bin/dwarfdump.rs:
```diff
@@ -2266,3 +2281,85 @@ fn dump_addr<R: Reader, W: Write>(w: &mut W, debug_addr: &gimli::DebugAddr<R>) -
+) -> Result<()> {
+    match macinfo {
+        gimli::DebugMacInfoItem::Define {
+            line: line_number,
+            text: macro_text,
```
```suggestion
            line,
            text,
```

Similar elsewhere.

DanielT commented:
>            text: macro_text,
Oof. That's the result of trying to automatically rename through rust-language-server, and I didn't check the diff afterwards.
Sorry for that!

It's fixed now.

philipc commented on src/read/macros.rs:
```diff
@@ -0,0 +1,283 @@
+                let macro_text = self.input.read_null_terminated_slice()?;
+                Ok(Some(DebugMacInfoItem::Define {
+                    line: line_number,
+                    text: macro_text,
```
Some more here still.

philipc reviewed (approved):
Thanks!
